### ISSUE TITLE
Bug: App Cleaner problem on unrooted device

### ISSUE THREAD

**Issue by Holguitar** [d4rken/sdmaid-public#4689]
![Screenshot_2021-03-08-15-56-23-716_com android settings](https://user-images.githubusercontent.com/18100580/110338111-ee4bf500-8026-11eb-98aa-9733eb9183a7.jpg)
![Screenshot_2021-03-08-15-22-06-436_eu thedarken sdm](https://user-images.githubusercontent.com/18100580/110338130-f5730300-8026-11eb-8f4f-ef110cd334f4.jpg)

### What happened?
On my unrooted Xiaomi Mi 9T running stock android 10 it's not possible to clean app caches autonated by app cleaner more than one time after installation even when accessibility service is activated. Other things get deleted/cleaned.
The second time I get a message that something with the accessibility service settings is not ok and I should try again.
Only way to make app cleaner work again completly is to delete data/caches of SD Maid. In this case it's working one time, after that I have to delete data again and re-activate all related settings.

### What did you expect to happen?
When accessibility service is enabled and SD Maid is on the list of the allowed apps I expect SD Maid to do its job by pressing the "clean caches" button of the related apps automated as fast as the system allows it every time I use the app cleaner.

### How can I reproduce the issue?
Install SD Maid on an unrooted device, activate accessibility service, give permissions in device settings and put SD Maid on the list of the allowed apps. Use app cleaner once (should work) and try it again (shouldn't work this time until you delete SD Maids data and cache).

### SD Maid
Version: `5.0.11 (50011)`
Pro: `isPro=true || isComplete=true (unl=true, iap=false, acc=false)`

### Device
Root: `RootContext(rootState=Root(state=UNAVAILABLE)`
Fingerprint: `Xiaomi/davinci_eea/davinci:10/QKQ1.190825.002/V12.0.4.0.QFJEUXM:user/release-keys`

**Comment by very-stack** [comment]
Wow

**Comment by chris-tian78** [comment]
same problem on poco m3 with miui 12

**Comment by webvan1999** [comment]
Same problem, AppCleaner doesn't do anything on my unrooted Mi 8.

**Comment by d4rken** [comment]
Sorry but this is not something I can fix.
It's related to your ROMs killing SD Maid and it's service in one way or another, usually as part of some "battery optimization" mechanism. Accessibility services can't be started by SD Maid, the system has to start them.

https://github.com/d4rken/sdmaid-public/wiki/AppCleaner#troubleshooting

**Comment by Holguitar** [comment]
Thanks for repliying and making that clear @d4rken. As the users that replied on my report were Xiaomi users too I expected something like that.
Sadly, the above workaround by @JeffHorinzon doesn´t seem to work for me on newer MIUI versions (MIUI 11, Android 10).

**Comment by JeffHorinzon** [comment]
> Thanks for repliying and making that clear @d4rken. As the users that replied on my report were Xiaomi users too I expected something like that.
> 
> Sadly, the above workaround by @JeffHorinzon doesn´t seem to work for me on newer MIUI versions (MIUI 11, Android 10).

Hi, there.  I just want to make sure.  Because the workaround I've found is using Redmi Note 7 (MIUI 12.0.1, Android 10).  As of writing, AppCleaner is cleaning as expected.  Perhaps you might've missed a step or two?  Do make sure Autostart is enabled for SDMaid and reboot the device after enabling the Accessibility service.

Should it still behave otherwise, please do record a debug log and reply to this thread.  Either Me or Darken will have a look at it and try to solve this issue :3

Thanks,
A friendly volunteer.

**Comment by C-M-D** [comment]
C'est pas moi qui pourrait t'aider car "perso" suis sur miui 12

Le sam. 13 mars 2021 à 21:05, Guard of Magog ***@***.***> a
écrit :

**Comment by Pablo403-60** [comment]
all of u guys i want to say thank you very very much, specially to mr. Matthias (d4rken) but unfortunetly this psychopath who is even playing on the screen im coming in hes out Im scrolling right hes left still nearly 3months no gmail who last more like 1 day he is sick calling me today and tell me ,,,,younges is better like oldest after i slept!!!,,@@@!!! with them......this idiot talking about my 12 yr. doughter and her mum 2days earlier whats happen when somebody meet my12yr. on the street!!!!?????!!!!#####----if thats hacking is a crazy scary thing then..Me cant do anymore im geting crazy for real anything what im trying to do he stop or change or dont let me do!!!!  exam. i order new bank card his order in system 4more and 2weeks gone im still waiting!!......icant do nothing on my phone cos he blocked my mi max2 xiaomi account again last time cost me 95pounds just take password out i dont have mony pay again i call xiaomi today ask for help ineed wait for email maybe i manage unblocked this next week then i try do what u guys sugest and show me im getting crazy he even change my sim mobile cards numbers!!!!! what!!!!!! and when i calling provider they dont belive me im on the bring i am traped and i scare whats happen with me someone need stop him cos i just cant my health is bad now..!! #™ thanks guys , Pablo_@.

**Comment by webvan1999** [comment]
> 
> 
> > Thanks for repliying and making that clear @d4rken. As the users that replied on my report were Xiaomi users too I expected something like that.
> > Sadly, the above workaround by @JeffHorinzon doesn´t seem to work for me on newer MIUI versions (MIUI 11, Android 10).
> 
> Hi, there. I just want to make sure. Because the workaround I've found is using Redmi Note 7 (MIUI 12.0.1, Android 10). As of writing, AppCleaner is cleaning as expected. Perhaps you might've missed a step or two? Do make sure Autostart is enabled for SDMaid and reboot the device after enabling the Accessibility service.
> 
> Should it still behave otherwise, please do record a debug log and reply to this thread. Either Me or Darken will have a look at it and try to solve this issue :3
> 
> Thanks,
> A friendly volunteer.

So I tried that but no go. There must be something with the MIUI 12 on the Mi 8 because in CCleaner I can't even see the "hidden cache" selection.
Fix for now is to run SDMain to identify biggest hidden caches and then delete them manually by going to the App details.

**Comment by d4rken** [comment]
@Holguitar One more thought.

> Use app cleaner once (should work) and try it again (shouldn't work this time until you delete SD Maids data and cache).

Can you elaborate on the timeframe here? By "once" you mean running it once, then hours later trying again? If you select 3 apps, clean them, select another 3 apps, it keeps working right?

So the procedure itself, can start multiple times directly in succession?

How long until it no longer works, and can this be speed up by force stopping SD Maid or rebooting the device?

**Comment by Holguitar** [comment]
>> Hi, there. I just want to make sure. Because the workaround I've found is using Redmi Note 7 (MIUI 12.0.1, Android 10). As of writing, AppCleaner is cleaning as expected. Perhaps you might've missed a step or two? Do make sure Autostart is enabled for SDMaid and reboot the device after enabling the Accessibility service.
> 
> Should it still behave otherwise, please do record a debug log and reply to this thread. Either Me or Darken will have a look at it and try to solve this issue :3
> 
> Thanks,
> A friendly volunteer.



> @Holguitar One more thought.
> 
> > Use app cleaner once (should work) and try it again (shouldn't work this time until you delete SD Maids data and cache).
> 
> Can you elaborate on the timeframe here? By "once" you mean running it once, then hours later trying again? If you select 3 apps, clean them, select another 3 apps, it keeps working right?
> 
> So the procedure itself, can start multiple times directly in succession?
> 
> How long until it no longer works, and can this be speed up by force stopping SD Maid or rebooting the device?

@d4rken @JeffHorinzon :
Thank you both for helping me out with that issue.
I updated to MIUI 12.0.4 eea now but that doesn't change anything. Nevertheless I got it working now. @JeffHorinzon was right, I missed a tiny but
apparently important step. I had to disable accessibility service (it was enabled in my case) **before** activating autostart for SD Maid, enable it again and rebooted.
So the step I missed was activating accessibility service before rebooting because it was already enabled.
So the steps in the correct order are:

**
- Disable accessibility service if enabled
- Enable autostart for SD Maid
- Enable accessibility service
- Reboot

That's it.**

@d4rken : By "once" I meant appcleaner does it's job exactly one time after a fresh install of SD Maid or optional after cleaning SD Maid's data/cache.
To make it fully working again you have to uninstall/reinstall SD Maid or clean data/cache again (which means you have to adjust all your settings again every time). So the timeframe doesn't matter at all.
Sorry if I wasn't clear in that case.
Anyway, thanks to @JeffHorinzon's workaround that procedure isn't needed anymore.
Maybe nevertheless it's time to root that Mi 9T too, it's the only one of all my past and present devices that is still unrooted....
Thank you all.

**Comment by JeffHorinzon** [comment]
You're welcome, I guess we need to update and clarify on the MIUI fix then.

@d4rken please check above comment.

**Comment by webvan1999** [comment]
Sorry to report that the fix is not working on my Mi 8 with MIUI 12.
When I hit "delete" I see the app details page in the background, 3x blinking and back to SDMaid and nothing erased. Oh and for good measure I had deactivated any "battery optimizations" for SDMaid too.

**Comment by d4rken** [comment]
> By "once" I meant appcleaner does it's job exactly one time after a fresh install of SD Maid or optional after cleaning SD Maid's data/cache.
> To make it fully working again you have to uninstall/reinstall SD Maid or clean data/cache again (which means you have to adjust all your settings again every time). So the timeframe doesn't matter at all.

Please be more specific. What do you mean by once? Once at all? Once if you closed SD Maid?
You click a single app in SD Maid, ACS clicks clear cache for you, you are returned to SD Maid, you click another app, and it no longer works?

> I updated to MIUI 12.0.4 eea now but that doesn't change anything. Nevertheless I got it working now. @JeffHorinzon was right, I missed a tiny but
apparently important step. I had to disable accessibility service (it was enabled in my case) before activating autostart for SD Maid, enable it again and rebooted.
So the step I missed was activating accessibility service before rebooting because it was already enabled.

I will update the wiki, thanks.

**Comment by d4rken** [comment]
> Sorry to report that the fix is not working on my Mi 8 with MIUI 12.
> When I hit "delete" I see the app details page in the background, 3x blinking and back to SDMaid and nothing erased. Oh and for good measure I had deactivated any "battery optimizations" for SDMaid too.

That's a different issue than the one @Holguitar  has. The issue here is that the ACS is not running, in your case it's running as you see something happening in the background. Your issue sounds like a language/ROM compatbility issue, open a new ticket and provide a debug log.

**Comment by maganezf** [comment]
> 

Thanks bro!! It works!!

**Comment by maganezf** [comment]
> Sorry but this is not something I can fix. It's related to your ROMs killing SD Maid and it's service in one way or another, usually as part of some "battery optimization" mechanism. Accessibility services can't be started by SD Maid, the system has to start them.
> 
> https://github.com/d4rken/sdmaid-public/wiki/AppCleaner#troubleshooting

Thanks bro!! It works!
